pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -81,6 +161,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 	return saClient, nil
 }
 
+func parseModels(annotations map[string]string) map[string]model {
+	models := map[string]model{}
+	for key, value := range annotations {
+		if prefix, name, ok := parseModelPrefixName(key); ok {
+			m := models[name]
+			switch prefix {
+			case OAuthRedirectModelAnnotationURISchemePrefix:
+				m.scheme = value
+			case OAuthRedirectModelAnnotationURIPortPrefix:
+				m.port = value
+			case OAuthRedirectModelAnnotationURIPathPrefix:
+				m.path = value
+			case OAuthRedirectModelAnnotationResourceKindPrefix:
+				m.kind = value
+			case OAuthRedirectModelAnnotationResourceNamePrefix:
+				m.name = value
+			case OAuthRedirectModelAnnotationResourceGroupPrefix:
+				m.group = value
+			}
+			models[name] = m
+		}
+	}
+	return models
+}
+
+func parseModelPrefixName(key string) (string, string, bool) {
+	for _, prefix := range modelPrefixes {
+		if strings.HasPrefix(key, prefix) {
+			return prefix, key[len(prefix):], true
+		}
+	}
+	return "", "", false
+}
+
+func extractValidRedirectURIs(models map[string]model, routeInterface osclient.RouteInterface) []redirectURI {
+	var data []redirectURI
+	var osRoutes []model
+
+	for _, model := range models {
+		if isOpenShiftRoute(&model) {
+			osRoutes = append(osRoutes, model)
+		}
+	}
+	if len(osRoutes) > 0 {
+		data = append(data, getOSRoutes(osRoutes, routeInterface)...)
+	}
+
+	return data
+}
+
+func getOSRoutes(modelList []model, routeInterface osclient.RouteInterface) []redirectURI {
+	var data []redirectURI
+	rm := map[string][]redirectURI{}
+
+	if len(modelList) > 1 {
+		routes, err := routeInterface.List(kapi.ListOptions{})
+		if err != nil {
+			return data
+		}
+		for _, r := range routes.Items {
+			updateRouteMap(rm, &r)
+		}
+	} else {
+		r, err := routeInterface.Get(modelList[0].name)
+		if err != nil {
+			return data
+		}
+		updateRouteMap(rm, r)
+	}
+
+	for _, m := range modelList {
+		if r, ok := rm[m.name]; ok {
+			for _, rURI := range r {
+				u := rURI
+				u.merge(m)
+				data = append(data, u)
+			}
+		}
+	}
+
+	return data
+}
+
+func updateRouteMap(rm map[string][]redirectURI, r *routeapi.Route) {
+	for _, i := range r.Status.Ingress {
+		if i.Host == "" {
+			continue
+		}
+		u := redirectURI{scheme: "https"}
+		u.host = i.Host
+		u.path = r.Spec.Path
+		u.port = r.Spec.Port.TargetPort.StrVal
+		rm[r.Name] = append(rm[r.Name], u)
+	}
+}
+
+func isOpenShiftRoute(m *model) bool {
+	return m.kind == RouteKind && m.group == routeapi.FutureGroupName
```

Comment:
> @deads2k is using FutureGroupName kosher?

It probably won't do what you want.  Is someone using it elsewhere?  I probably wouldn't do that yet.